lib/spack/spack/cmd/solve.py

```diff
@@ -102,21 +104,29 @@ def solve(parser, args):
 
     specs = spack.cmd.parse_specs(args.specs)
 
-    # dump generated ASP program
-    result = asp.solve(
-        specs, dump=dump, models=models, timers=args.timers, stats=args.stats,
-        reuse=args.reuse,
+    # set up solver parameters
+    # Note: reuse and other concretizer prefs are passed as configuration
+    solver = asp.Solver()
+    output = sys.stdout if "asp" in show else None
+    result = solver.solve(
+        specs,
+        out=output,
+        models=models,
+        timers=args.timers,
+        stats=args.stats,
+        setup_only=not any(s in show for s in ("output", "solutions"))
```

Comment:
you need to run the solve to show `opt` too

```suggestion
        setup_only=(set(show) != {'asp'})
```